index.bs

```diff
@@ -2129,11 +2189,103 @@ an [=attribution trigger=] |trigger|:
+    :: |reportTime|
+    : [=aggregatable report/report time=]
+    :: |reportTime|
+    : [=aggregatable report/report id=]
+    :: The result of [=generating a random UUID=]
+    : [=aggregatable report/source debug key=]
+    :: null
+    : [=aggregatable report/trigger debug key=]
+    :: |trigger|'s [=attribution trigger/debug key=]
+    : [=aggregatable report/contributions=]
+    :: « |contribution| »
+    : [=aggregatable report/serialized private state token=]
+    :: null
     : [=aggregatable report/aggregation coordinator=]
     :: |trigger|'s [=attribution trigger/aggregation coordinator=]
+    : [=aggregatable report/source registration time configuration=]
+    :: |trigger|'s [=attribution trigger/aggregatable source registration time configuration=]
+    : [=aggregatable report/is null report=]
+    :: true
 1. Return |report|.
 
+To <dfn>generate null reports</dfn> given an [=attribution trigger=] |trigger|:
+
+1. Let |nullReports| be a new [=list/is empty|empty=] [=list=].
+1. If |trigger|'s [=attribution trigger/aggregatable source registration time configuration=] is "<code>[=aggregatable source registration time configuration/exclude=]</code>":
+    1. Let |numNullReports| be the result of [=obtaining a randomized number of null reports=] with
+        [=randomized null report distribution parameter excluding source registration time=].
+    1. Repeat the following steps |numNullReports| times:
+        1. Let |nullReport| be the result of [=obtaining a null report=] with |trigger| and
+            |trigger|'s [=attribution trigger/trigger time=].
+        1. Add |nullReport| to the [=aggregatable report cache=].
+        1. [=list/Append=] |nullReport| to |nullReports|.
+    1. Return |nullReports|.
+1. Let |maxSourceExpiry| be [=max source expiry=].
+1. Round |maxSourceExpiry| away from zero to the nearest day (86400 seconds).
+1. For each integer |day| between 0 (inclusive) and the number of days in |maxSourceExpiry| (inclusive):
+    1. Let |numNullReports| be the result of [=obtaining a randomized number of null reports=] with
+        [=randomized null report distribution parameter including source registration time=].
+    1. Repeat the following steps |numNullReports| times:
+        1. Let |fakeSourceTime| be |trigger|'s [=attribution trigger/trigger time=] - |day| days.
+        1. Let |nullReport| be the result of [=obtaining a null report=] with |trigger| and |fakeSourceTime|.
+        1. Add |nullReport| to the [=aggregatable report cache=].
+        1. [=list/Append=] |nullReport| to |nullReports|.
+1. Return |nullReports|.
+
+<h3 id="assign-private-state-tokens-for-aggregatable-reports">Assigning private state tokens for aggregatable reports</h3>
+
+To <dfn>shuffle a [=list=]</dfn> |list|, reorder |list|'s elements such that each possible permutation has equal
+probability of appearance.
+
+To <dfn>assign private state tokens</dfn> given a [=list=] of [=aggregatable reports=] |reports| and
```

Comment:
This algorithm also has to assign the proper `report id` to the null reports, otherwise the token will have a mismatched ID.

Given that we don't currently specify how the report ID gets in the token, I think we can just punt this to an ISSUE, but we should document the problem.

Reply:
Yeah I left this out as report ID assignment is not spec'ed yet. Added an issue for that.